dali/python/nvidia/dali/external_source.py

```diff
@@ -97,32 +100,47 @@ def __init__(
     def append(self, instance):
         self.instances.append(instance)
 
-    def callback_args(self, idx_in_batch):
+    def callback_args(self, idx_in_batch, batch_size = 0, lead = 0):
         """Generate information to be passed to ES callback.
 
         Args:
             idx_in_batch: Index in batch for per-sample mode, None indicates batch mode where we
             pass only the iteration number.
+            lead: how many batches ahead is this job wrt actual iteration
         """
         if not self.accepts_arg:
             return ()
         if idx_in_batch is not None:
-            arg = nvidia.dali.types.SampleInfo(self.current_sample + idx_in_batch, idx_in_batch, self.current_iter)
+            arg = nvidia.dali.types.SampleInfo(
+                self.current_sample + idx_in_batch + batch_size * lead,
+                idx_in_batch,
+                self.current_iter + lead)
         else:
-            arg = self.current_iter
+            arg = self.current_iter + lead
         return (arg,)
 
     def reset_indices(self):
         self.current_iter = 0
         self.current_sample = 0
+        self.cancel_prefetch()
 
-    def schedule_batch(self, pool, context_i, batch_size):
+    def cancel_prefetch(self):
+        self.scheduled_ahead = 0
```

Comment:
Does that cancel prefetch (I don't see how `self.scheduled_ahead = 0` does that)?

Reply:
This is sort of spread between pool.reset and this place. Pool.reset marks the outputs as ignored, whereas this makes it possible to reschedule the batches.

Reply:
What if someone calls `cancel_prefetch` while `schedule_and_feed` is running? You can get `self.scheduled_ahead==-1`